### PR TITLE
add lendex spec

### PR DESCRIPTION
**Pull request by Benjmhart**

### CODE REVIEW

File: mlabs/lendex-endpoint-spec.md [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+
+### AddReserve
+
+input: Mlabs.Lending.Contract.Api.AddReserve
+
+Prerequisite: none
+
+behaviour: 
+
+-- @anton:  what does this endpoint do?
```
Comment (anton-k) [inline mlabs/lendex-endpoint-spec.md:58]: It initializes reserve for given currencies.

File: mlabs/lendex-endpoint-spec.md [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+
+if repaying Ada, the user's stake delegation should be changed to some default? - if this feature is available in plutus
+
+invariant outcomes:
+can't repay a negative amount but 0 is ok.
+
+### SwapBorrowRateModel
+
+### LiquidationCall
+
```
Comment (anton-k) [inline mlabs/lendex-endpoint-spec.md:204]: Liquidation call is described here (it's implemented by this spec)

https://docs.aave.com/faq/liquidations

File: mlabs/lendex-endpoint-spec.md [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+All Lendex endpoints rely on the Lendex being initialized through a call to `StartParams`
+
+### Deposit
+
+input:
+Mlabs.Lending.Contract.Api.Deposit (amount, asset)
+
+Prerequisite: wallet holds underlying currency for Market
+
+Expected outcome:
```
[the remark this answers is not in the document]
Comment (anton-k) [reply mlabs/lendex-endpoint-spec.md:84]: Also it's good tosay that we mint new aTokens as a result of this TX

File: mlabs/lendex-endpoint-spec.md [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+input:
+Mlabs.Lending.Contract.Api.Withdraw (amount, asset)
+
+Prerequisite: The user must hold aTokens for the market in question, making the `Deposit` endpoint necessary
+
+
+Expected outcome:
+Wallet Balance of aToken is reduced by x.
+
+Wallet balance of underlying is increased by (x * e) (less fees)
```
[the remark this answers is not in the document]
Comment (anton-k) [reply mlabs/lendex-endpoint-spec.md:107]: Also it's good to say that we burn aTokens as a result of this TX. This is one of the effect (see the list of effects of this action).

File: mlabs/lendex-endpoint-spec.md [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+the wallet of the user will decrease by the specified amount from input in the Market's underlying Currency (or round down if the amount available is less than the amount in the input)
+the outstanding loan payment will be reduced by the amount paid, less interest applied against the loan
+Collateral is not released on this endpoint, even if the borrow amount is totally paid off.
+
+if repaying Ada, the user's stake delegation should be changed to some default? - if this feature is available in plutus
+
+invariant outcomes:
+can't repay a negative amount but 0 is ok.
+
+### SwapBorrowRateModel
```
Comment (anton-k) [inline mlabs/lendex-endpoint-spec.md:201]: I guess it's not implemented yet. In Aave it allows to switch between Stable and Variable Rate models.
Those models differ in quotients of the transition function.

File: mlabs/nft-endpoint-spec.md [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+### SetPrice
+
+prerequiste: none beyond contract instantiation
+must be the current owner
+
+input:
+Mlabs.Nft.Contract.Api.SetPrice
+
+behavior:  
+updates the `price` parameter needed for the `Buy` endpoint
```
Comment (anton-k) [inline mlabs/nft-endpoint-spec.md:35]: Maybe it's worth to say that we can lock the NFT by setting Nothing as a price

File: mlabs/nft-endpoint-spec.md [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+the current asking price specified by a call to either `StartParams` or `SetPrice` must be a Just.   if it is a Nothing, then the asset is not for sale.
+
+input:
+Mlabs.Nft.Contract.Api.Buy
+(price, newprice)
+
+behavior:
+
+if the Buy.price is greater than or equal to the asking price, the user's wallet will be reduced by Buy.Price Ada (the contract must fail if the user has less than the specified Buy.price)
+the funds sent by the caller are split such that the `share` parameter amount is sent to the author, and the remainder is sent to the current owner
```
Comment (anton-k) [inline mlabs/nft-endpoint-spec.md:49]: Share is not a fixed amount it's portion of the current buy price

Comment (anton-k) [reply mlabs/nft-endpoint-spec.md:49]: It's not like that. Buyer pays amount that buyer has specified to the current owner and portion of shar eto the author.
we do not pay share and all the rtest that the buyer specified. Buyer should provide so many coins as buyer has specified in the amount parameter of the input plus share portion.

**Review by anton-k — CHANGES_REQUESTED** [review]

File: mlabs/lendex-endpoint-spec.md (hunk 1 above)
Comment (Benjmhart) [reply mlabs/lendex-endpoint-spec.md:58]: maybe my work on other lendex platforms has confused me, can you elaborate on where the reserve is used in the system? or is it just a safety of some kind?

Comment (Benjmhart) [reply mlabs/lendex-endpoint-spec.md:58]: is the reserve in the aToken or the underlying?

File: mlabs/lendex-endpoint-spec.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (Benjmhart) [reply mlabs/lendex-endpoint-spec.md:84]: can you clarify a bit?  it seems like this is related to the reserve? 

i will definitely add that we mint aTokens here. thanks.

[the remark this answers is not in the document]
Comment (Benjmhart) [reply mlabs/lendex-endpoint-spec.md:84]: or is it that we don't round down and the request just fails if there isn't enough funds to match the specified amount?

File: mlabs/lendex-endpoint-spec.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (Benjmhart) [reply mlabs/lendex-endpoint-spec.md:107]: we say that we burn aTokens on the next line.

is this also about rounding or some other inaccuracy?

File: mlabs/lendex-endpoint-spec.md [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+
+collateral ratios are calculated using each aToken : underlying exchange rate, then the exchange rate between that underlying token and Ada, which is provided by the Oracle - this way the collateral and the borrowed currency can be compared via relative value in Ada.
+
+behavior:
+let `userTokens` equal amount of `asset` in user wallet / provided utxos
+let `contractTokens` equal amount of `asset` currently locked as collateral for this user
+if useAsCollateral is True, then move (`userTokens` of `asset` * `portion`) from User wallet to contract and lock as collateral
+if useAsCollateral is false, then move (`contractTokens` of `asset` * `portion`) from contract to user wallet rounding down to ensure that the user does not go below minimum collateral ratios for this Lendex/User contract
+
+@anton - why is the `useAsCollateral` flag necessary? the Api for this endpoint feels a bit strange
```
[the remark this answers is not in the document]
Comment (Benjmhart) [reply mlabs/lendex-endpoint-spec.md:142]: is the behavior described correct?

File: mlabs/lendex-endpoint-spec.md (hunk 5 above)
Comment (Benjmhart) [reply mlabs/lendex-endpoint-spec.md:201]: can you describe what Stable and Variable rates mean? is this documented elsewhere?

File: mlabs/lendex-endpoint-spec.md (hunk 2 above)
Comment (Benjmhart) [reply mlabs/lendex-endpoint-spec.md:204]: thanks.

File: mlabs/nft-endpoint-spec.md (hunk 6 above)
Comment (Benjmhart) [reply mlabs/nft-endpoint-spec.md:35]: sure.

File: mlabs/nft-endpoint-spec.md (hunk 7 above)
Comment (Benjmhart) [reply mlabs/nft-endpoint-spec.md:49]: i think i meant this, but it's not very clear the way it was written, thanks for the heads up

File: mlabs/lendex-endpoint-spec.md (hunk 1 above)
Comment (anton-k) [reply mlabs/lendex-endpoint-spec.md:58]: Reserve keeps all internal wallets and accounts. How much liquidity we have a lot of things.

see https://github.com/mlabs-haskell/plutus-use-cases/blob/main/mlabs/src/Mlabs/Lending/Logic/Types.hs#L91

File: mlabs/lendex-endpoint-spec.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (anton-k) [reply mlabs/lendex-endpoint-spec.md:92]: You say that we convert something with exchange rate. But in fact we jast add values to internal balances or substract them on withdraw.

File: mlabs/lendex-endpoint-spec.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (anton-k) [reply mlabs/lendex-endpoint-spec.md:153]: Yes, it is

File: mlabs/lendex-endpoint-spec.md (hunk 1 above)
Comment (Benjmhart) [reply mlabs/lendex-endpoint-spec.md:58]: @anton-k - i'm afraid i'm still not quite sure why this is necessary, does a user also deposit funds to a reserve, which are used for loans? how is this different than regular deposits?

Comment (anton-k) [reply mlabs/lendex-endpoint-spec.md:58]: It initializes the reserve. We can init the reserve and give funds to it. On the plutus level we keep all deposits attached to the same script. But we keep accounting which funds belong to which reserve in the reserve's datum.

File: mlabs/lendex-endpoint-spec.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (anton-k) [reply mlabs/lendex-endpoint-spec.md:115]: I mean that funds that are deposited are added to the reserve funds, we do not convert them in any way,no  need to scale here.

**Review by anton-k — COMMENTED** [review]
Added comments on unresolved topics

**Pull request merged**